sdk/identity/identity/src/credentials/msalBrowser/msalImplicit.ts

```diff
@@ -0,0 +1,230 @@
+import { AuthenticationRecord } from "../../client/msalClient";
+import { credentialLogger } from "../../util/logging";
+import {
+  BrowserLoginStyle,
+  InteractiveBrowserAuthenticateOptions
+} from "../interactiveBrowserCredentialOptions";
+import { IMSALBrowserFlow, IMSALToken, MSALOptions } from "./msalCommon";
+
+const logger = credentialLogger("MSAL Browser v1 - Implicit Grant Flow");
+
+/**
+ * Uses MSAL directly for browser authentication,
+ * which uses the Implicit Grant Flow:
+ * <https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-implicit-grant-flow>
+ */
+export class MSALImplicit implements IMSALBrowserFlow {
+  private config: msal.Configuration;
+  private app: msal.UserAgentApplication;
+  private loginStyle: BrowserLoginStyle;
+  private correlationId?: string;
+  private tenantId: string;
+  private account: AuthenticationRecord | undefined;
+
+  /**
+   * Sets up an MSAL object based on the given parameters.
+   * MSAL with Implicit Grant is not compatible with the account workflow we use from the MSAL Auth Code flow.
+   * In this case, any `authenticationRecord` received will be ignored.
+   * @param options - Parameters necessary and otherwise used to create the MSAL object.
+   */
+  constructor(options: MSALOptions) {
+    this.loginStyle = options.loginStyle;
+    this.correlationId = options.correlationId;
+    this.tenantId = options.tenantId!;
+    this.config = {
+      auth: {
+        clientId: options.clientId!, // we just initialized it above
+        authority: `${options.authorityHost}/${options.tenantId}`,
+        knownAuthorities: options.knownAuthorities,
+        // If the users picked redirect as their login style,
+        // but they didn't provide a redirectUri,
+        // we can try to use the current page we're in as a default value.
+        redirectUri: options.redirectUri || window.location.origin,
+        postLogoutRedirectUri: options.postLogoutRedirectUri
+      },
+      cache: {
+        cacheLocation: "localStorage",
+        storeAuthStateInCookie: true // Set to true to improve the experience on IE11 and Edge.
+      }
+    };
+
+    this.app = new msal.UserAgentApplication(this.config);
+  }
+
+  /**
+   * Formats an MSAL 1 account into an `AuthenticationRecord`.
+   * @param account - The account in the shape defined by MSAL.
+   */
+  private handleAccount(account: msal.Account): AuthenticationRecord {
+    return {
+      homeAccountId: account.homeAccountIdentifier,
+      environment: account.environment,
+      tenantId: this.tenantId,
+      localAccountId: account.accountIdentifier,
+      username: account.environment
+    };
+  }
+
+  /**
+   * Loads the account based on the result of the authentication.
+   * If no result was received, tries to load the account from the cache.
+   * @param result - Result object received from MSAL.
+   */
+  private handleResult(result: msal.AuthResponse | null): AuthenticationRecord | undefined {
```

Comment:
for some reason this read funny to me. Is the `msal.AuthResponse | null` and expected value?

Reply:
Thanks! I'll ensure it's only authResponse

Reply:
I'll make it `result?: msal.AuthResponse`. How does that sound?

Reply:
Oh this might help! Some methods that retrieve a result from MSAL can return "null", so that's why I had this `| null`. I think it's clearer if I handle the null into an undefined, so in those cases now I use something like `(await msal.getResult()) || undefined`. Thoughts appreciated!